fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/OrToUnionAllJoinRule.java

```diff
@@ -0,0 +1,346 @@
+            childOutputColumns.add(outputColumns);
+        }
+
+        LogicalUnionOperator unionOp =
+                new LogicalUnionOperator.Builder().isUnionAll(true).setProjection(joinOp.getProjection())
+                        .setChildOutputColumns(childOutputColumns).setOutputColumnRefOp(outputColumns).build();
+
+        OptExpression result = OptExpression.create(unionOp, unionChildren);
+        return Lists.newArrayList(result);
+    }
+
+    private boolean containsLimitOrNonDeterministicFunction(OptExpression expr) {
+        if (expr.getOp().getOpType() == OperatorType.LOGICAL_LIMIT) {
+            return true;
+        }
+
+        if (expr.getOp() instanceof LogicalOperator) {
+            LogicalOperator logicalOp = (LogicalOperator) expr.getOp();
+
+            if (logicalOp.getPredicate() != null &&
+                    containsNonDeterministicFn(logicalOp.getPredicate())) {
+                return true;
+            }
+
+            if (logicalOp.getProjection() != null) {
+                for (ScalarOperator projExpr : logicalOp.getProjection().getColumnRefMap().values()) {
+                    if (containsNonDeterministicFn(projExpr)) {
+                        return true;
+                    }
+                }
+            }
+        }
+
+        for (OptExpression child : expr.getInputs()) {
+            if (containsLimitOrNonDeterministicFunction(child)) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    private boolean containsNonDeterministicFn(ScalarOperator expr) {
+        if (expr instanceof CallOperator) {
+            CallOperator callOp = (CallOperator) expr;
+            String functionName = callOp.getFnName();
+            if (functionName != null &&
+                    FunctionSet.allNonDeterministicFunctions.contains(functionName)) {
+                return true;
+            }
+        }
+
+        for (ScalarOperator child : expr.getChildren()) {
+            if (containsNonDeterministicFn(child)) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    private Map<Integer, List<BinaryPredicateOperator>> createCumulativePredicateMap(
+            List<BinaryPredicateOperator> binaryPredicateList) {
+
+        Map<Integer, List<BinaryPredicateOperator>> cumulativePredicateMap = new HashMap<>();
+
+        for (int i = 0; i < binaryPredicateList.size(); i++) {
+            List<BinaryPredicateOperator> currentBranchPredicates = new ArrayList<>();
+
+
+            for (int j = 0; j <= i; j++) {
+                BinaryPredicateOperator originalPredicate = binaryPredicateList.get(j);
+                Map<ColumnRefOperator, ColumnRefOperator> columnMapping = new HashMap<>();
+                collectColumnRefs(originalPredicate, columnMapping);
+
+                ReplaceColumnRefRewriter rewriter = new ReplaceColumnRefRewriter(columnMapping);
```

Comment:
it will use the original columnRef Id in this block. you can refer to `OptExpressionDuplicator`